### PR TITLE
Add retry policies into DPS device

### PR DESCRIPTION
**Pull request by brycewang-microsoft**
Will have another PR to add retry policies for DPS **service**.

### CODE REVIEW

File: iothub/service/tests/RetryPolicies/IotHubServiceRetryPolicyBaseTests.cs [hunk 1]
```diff
@@ -45,7 +45,7 @@ public void RetryPolicyBase_ObservesInifiniteRetries()
         [TestMethod]
         [DataRow(true)]
         [DataRow(false)]
-        public void RetryPolicyBase_IotHubException_ReturnsTrueWhenTransient(bool isTransient)
+        public void RetryPolicyBase_IotHubServiceException_ReturnsTrueWhenTransient(bool isTransient)
```
Comment (brycewang-microsoft) [inline iothub/service/tests/RetryPolicies/IotHubServiceRetryPolicyBaseTests.cs:48]: correcting the test method name as we don't have `IotHubException` context any more.

File: SDK v2 migration guide.md [hunk 2]
```diff
@@ -328,6 +328,8 @@ What was a loose affiliation of separate clients is now a consolidated client wi
 - Added support for setting a client web socket instance in the client options so that users can have better control over AMQP web socket connections.
 - Added support for setting the web socket level keep alive interval for AMQP web socket connections.
 - Added support for setting the remote certificate validation callback for AMQP TCP connections.
+- The library now includes `IProvisioningClientRetryPolicy` implementations: `ProvisioningClientExponentialBackoffRetryPolicy`, `ProvisioningClientFixedDelayRetryPolicy`, `IotHubServiceIncrementalDelayRetryPolicy` and `ProvisioningClientNoRetry`,
+ which can be set by calling `ProvisioningClientOptions.RetryPolicy`.
```
Comment (drwill-ms) [inline SDK v2 migration guide.md:332]:
```suggestion
 which can be set by setting `ProvisioningClientOptions.RetryPolicy`.
```

File: provisioning/device/src/RetryPolicies/ProvisioningClientExponentialBackoffRetryPolicy.cs [hunk 3]
```diff
@@ -0,0 +1,74 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Globalization;
+using System;
```
Comment (drwill-ms) [inline provisioning/device/src/RetryPolicies/ProvisioningClientExponentialBackoffRetryPolicy.cs:5]: sort

File: SDK v2 migration guide.md (hunk 2 above)
Comment (brycewang-microsoft) [reply SDK v2 migration guide.md:332]: "set by setting" sounds repetitive so I will say "which can be set via ..."

File: provisioning/device/src/RetryPolicies/ProvisioningClientFixedDelayRetryPolicy.cs [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        /// Creates an instance of this class.
+        /// </summary>
+        /// <param name="maxRetries">The maximum number of retry attempts; use 0 for infinite retries.</param>
+        /// <param name="fixedDelay">The fixed delay to wait between retries.</param>
+        /// <param name="useJitter">Whether to add a small, random adjustment to the retry delay to avoid synchronicity in retrying clients.</param>
+        /// <exception cref="ArgumentOutOfRangeException">Throw if the value of <paramref name="fixedDelay"/> is negative.</exception>
+        public ProvisioningClientFixedDelayRetryPolicy(uint maxRetries, TimeSpan fixedDelay, bool useJitter = true)
+            : base(maxRetries)
+        {
+            Argument.AssertNotNegativeValue(fixedDelay.Ticks, nameof(fixedDelay));
```
Comment (drwill-ms) [inline provisioning/device/src/RetryPolicies/ProvisioningClientFixedDelayRetryPolicy.cs:30]: Could this be changed/copied to use a TimeSpan parameter?

File: provisioning/device/src/RetryPolicies/ProvisioningClientRetryPolicyBase.cs [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        /// <summary>
+        /// Creates an instance of this class.
+        /// </summary>
+        /// <param name="maxRetries">The maximum number of retries; use a negative value for infinite retries.</param>
+        /// <exception cref="ArgumentOutOfRangeException">Throw if the value of <paramref name="maxRetries"/> is negative.</exception>
+        protected ProvisioningClientRetryPolicyBase(uint maxRetries)
+        {
+            if (maxRetries < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(maxRetries), maxRetries, string.Format(CultureInfo.CurrentCulture, "ArgumentCannotBeNegative", new object[]
```
Comment (drwill-ms) [inline provisioning/device/src/RetryPolicies/ProvisioningClientRetryPolicyBase.cs:26]: Needs updating.

**Review by drwill-ms — APPROVED** [review]

File: e2e/test/provisioning/ProvisioningE2ETests.cs [hunk 6]
```diff
@@ -858,8 +861,8 @@ public static ProvisioningClientOptions CreateProvisioningClientOptionsFromName(
             if (transportSettings is IotHubClientMqttSettings)
             {
                 return transportSettings.Protocol == IotHubClientTransportProtocol.Tcp
-                    ? new ProvisioningClientOptions(new ProvisioningClientAmqpSettings(ProvisioningClientTransportProtocol.Tcp))
```
Comment (brycewang-microsoft) [inline e2e/test/provisioning/ProvisioningE2ETests.cs:861]: this was wrong, so correct.

**Pull request merged**